### PR TITLE
test: Wrap environment with cloudprovider environment

### PR DESCRIPTION
**Pull request by jonathan-innis**
Fixes # 

**Description**

- Wraps/Decorates the standard `test/suites` `Environment` with `AWSEnvironment`
- Migrates all cloudprovider-specific APIs into a separate struct
- Separates the cloudprovider-specific expectations

**How was this change tested?**

* `TEST_FILTER=Integration make e2etests`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

**Release Note**

```release-note
None
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/pkg/environment/environment.go [hunk 1]
```diff
@@ -40,15 +41,37 @@ import (
 	"github.com/aws/karpenter/pkg/utils/project"
 )
 
+type AWSEnvironment struct {
```
Comment (ellistarn) [inline test/pkg/environment/environment.go:44]: Can we make this a new file? Sounds like we may want to port part of this to karpenter-core?

File: test/pkg/environment/setup.go [hunk 2]
```diff
@@ -209,7 +210,7 @@ func (env *Environment) startNodeMonitor(stop <-chan struct{}) {
 	factory.Start(stop)
 }
 
-func (env *Environment) AfterEach() {
+func (env *Environment) AfterEach(excludeObjs ...client.Object) {
```
Comment (ellistarn) [inline test/pkg/environment/setup.go:213]: Maybe worth passing in an options object env.Exclude(objects...) to make this interface more apparent. Right now, the caller can't really tell what this means.

**Review by ellistarn — APPROVED** [review]

File: test/pkg/environment/setup.go (hunk 2 above)
Comment (jonathan-innis) [reply test/pkg/environment/setup.go:213]: Sounds good 👍

**Review by ellistarn — APPROVED** [review]

File: test/pkg/environment/aws/setup.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		{&v1alpha1.AWSNodeTemplate{}, &v1alpha1.AWSNodeTemplateList{}},
+	}
+)
+
+func (env *Environment) BeforeEach(opts ...common.Option) {
+	options := common.ResolveOptions(opts)
+	if options.EnableDebug {
+		fmt.Println("------- START BEFORE -------")
+		defer fmt.Println("------- END BEFORE -------")
+	}
```
Comment (ellistarn) [inline test/pkg/environment/aws/setup.go:38]: Logging seems odd to be AWS specific.

File: test/pkg/environment/common/environment.go [hunk 4]
```diff
@@ -40,15 +35,16 @@ import (
 	"github.com/aws/karpenter/pkg/utils/project"
 )
 
+type Pair[A, B any] struct {
```
Comment (ellistarn) [inline test/pkg/environment/common/environment.go:38]: optional: toss this in `functional`

File: test/pkg/environment/common/setup.go [hunk 5]
```diff
@@ -321,3 +317,7 @@ func (env *Environment) dumpNodeEvents(testStartTime time.Time) {
 		fmt.Print(getEventInformation("node", k, v))
 	}
 }
+
+func containsObjectType(objs []client.Object, obj client.Object) bool {
```
Comment (ellistarn) [inline test/pkg/environment/common/setup.go:321]: optional: I typically making one line wrappers for higher order functions unless there's some real differentiating magic going on.

**Review by ellistarn — APPROVED** [review]
Very cool!

**Review by ellistarn — APPROVED** [review]

**Pull request merged**